packages/apollo-server-express/src/__tests__/ApolloServer.test.ts

```diff
@@ -38,6 +37,7 @@ describe('apollo-server-express', () => {
       server = new ApolloServer(options);
       const app = express();
       server.applyMiddleware({ app });
+      const port = await getPortPromise({ port: 6665 });
```

Comment:
What's the thought behind continuing to pass a minimum `port`?  Couldn't we just leave the `port` option off entirely?

Reply:
In theory yes. I will  But when I checkout the master branch (without my changes) and try to run `npm run test`, I got a few weird errors. Like this one:

```
● RESTDataSource › constructing requests › serializes a request body that is an object as JSON

    expect(received).toEqual(expected) // deep equality

    Expected: "{\"foo\":\"bar\"}"
    Received: {"data": [123, 34, 102, 111, 111, 34, 58, 34, 98, 97, 114, 34, 125], "type": "Buffer"}

      261 |       expect(fetch.mock.calls.length).toEqual(1);
      262 |       expect(fetch.mock.calls[0][0].url).toEqual('https://api.example.com/foo');
    > 263 |       expect(fetch.mock.calls[0][0].body).toEqual(
          |                                           ^
      264 |         JSON.stringify({ foo: 'bar' }),
      265 |       );
      266 |       expect(fetch.mock.calls[0][0].headers.get('Content-Type')).toEqual(

      at Object.<anonymous> (src/__tests__/RESTDataSource.test.ts:263:43)
      at fulfilled (src/__tests__/RESTDataSource.test.ts:4:58)
```

Any clue? I didn't find anything in the issue list

Reply:
@mondaychen Hmm, it did just work for me.  What version of Node.js are you running them with?  Could you try an `npm run clean` followed by an `npm cit` (`npm clean-install-test`)?

Reply:
I'm using v10.15.3. I actually fixed it by adding `.toString()` to `body`.  Now I'm looking at a timeout issue in `packages/apollo-server/src/__tests__/index.test.ts`. I'll try the commands you mentioned.

Which version are you using?

Reply:
I ran `npm run clean` and `npm cit` and got same result. Maybe it's a version thing?

Reply:
Hmm.  `v10.15.1`

Reply:
So I've fixed the problems I had. When I leave the ports entirely to getPortPromise, I have a few errors such as "socket hang up", which means port conflicting happens again. My best guess is that jest is trying to start some of the tests in parallel, so they find the same port (such as 8000) available.

Reply:
Jest certainly does run most of the tests in parallel.  I thought `portfinder` was aimed to solve this problem by atomically grabbing the port prior to returning its promise and trying another if it failed?  Perhaps there's a similar package that does this?

If not, one other option here might be to utilize `process.env.JEST_WORKER_ID` which is a 1-based incremental number for each Jest worker.  No more than one test will run at a time in a specific worker, but there can be dozens and dozens of workers, depending on your CPU core count (Jest takes 50%, I think).

Ideally, we could just not depend on these port numbers at all but a thought that might work for the foreseeable future (until we can quell this dependency on ports) might be to pass the following to `getPortPromise`:

```
port: parseInt(`8${process.env.JEST_WORKER_ID}00`),
```

Though `process.env.JEST_WORKER_ID` should be memoized as it's an expensive lookup.

Reply:
I think you are right about `portfinder`, but because the tests are started in parallel, a lot of `getPortPromise` are returned before any of the port registering (`server.listen`) happens. In other words, the port was available when `getPortPromise` was executed (so it didn't fail) but became unavailable later the tests.

I think `process.env.JEST_WORKER_ID` should work. But if it's expensive, maybe we can have a function inside `apollo-server-test-suite` to make sure we never use a "will-be-used" port during a test run.

Reply:
The only thing expensive with what I proposed, I think, is reading from `process.env`, which has a performance cost because it has to call into C++ code (via `getenv(3)`, I believe).

Part of securing a random port is ensuring that it's received, so if `portfinder` isn't doing this, regardless of race conditions, it's perhaps just unable to do its job because the port is getting free'd up and re-picked immediately by another worker.  

So perhaps what we need to do is just ask the kernel to assign a random port ourselves?  Node.js supports this natively (and I suspect Apollo Server, possibly out of the box?) [by passing `0` as the `port` to `Net.prototype.listen`](https://nodejs.org/api/net.html#net_server_listen_port_host_backlog_callback).

I'm not sure how much work it would take to make our tests happy with this pattern, but it should, conceivably, be possible!

Reply:
Sounds perfect. Will do.